README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# windows.pls

<!-- badges: start -->
<!-- badges: end -->

Since spectra have been used to investigate the content of a sample,
Chemometrics, the field which studies how mathematical and statistical
methods can be applied to Chemistry, has been developing several models
that are useful to predict concentration of substances of interest.

This technology is widely used because of the fact that it is quick and
green, which means that it does not impact the environment and also is
not time consuming. On the other hand, ad hoc methods are needed to
explore high dimensional data and Partial Least Squares Regression (PLS)
is one the most used models.

For years now, researchers have been questioning if spectra in its
entirety could be less useful than selecting only smaller regions which
might produce better predictions. From this, the intent of this package,
which aims at dividing spectra in windows, which are used in PLS as
predictors and then compared to the complete spectra, to see if there
are improvements. In this package there are three possible modes:

- **wpls**, which stands for *Windows Window PLS*, divides the original
  spectra into several windows, computes PLS and stores metrics of
  interest such as RMSE and R2 for calibration and cross-validation
  both.

- **ewpls**, which stands for *Evolving Window PLS*, divides the
  original spectra into several windows, but each new window
  incorporates the previous ones, so that we are comparing smaller
  windows with the entire spectra.

- **swpls**, which stands for *Sliding Window PLS*, asks the width of
  the window that will be used to compute the model and the step that
  the window will make forward in the spectra so that a new model is
  calculated. In this way the window slides along spectra and computes
  several models, which will be compared with metrics.

## Installation

You can install the development version of windows.pls like so:

``` r
install_github("egonzato/windows.pls")
```

## Example

This is a basic example which shows you how to use this package:

``` r
library(windows.pls)
data("beer")
conc=unlist(beer[,1])
spectra=beer[,2:ncol(beer)]
names(spectra)=convert.names.wl(1100,2250,2)
```

Let’s plot our specta mapping them with respect to the associated
concentration:

``` r
map.spectra.gradient(spectra,conc)
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

Now, let’s suppose we would like to test three different windows, how
would those look?

``` r
segment.windows(spectra, conc, windows = 3)
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />

Now, we should divide our dataset into training and testing:

``` r
set.seed(181299)
rows=sample(nrow(spectra),round(nrow(spectra)/3))
xtrain=spectra[-rows,]; ytrain=conc[-rows]
xtest=spectra[rows,]; ytest=conc[rows]
names(xtrain)=convert.names.wl(1100,2250,2)
names(xtest)=convert.names.wl(1100,2250,2)
```

Let’s compute a model with three windows:

``` r
mywpls=cv.wpls(xtrain,ytrain,mode = 'wpls', windows = 3)
```

Now, to compare the four possible models, we can plot RMSE of
calibration and cross-validation for each of them, in a single window:

``` r
global.rmse(mywpls)
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

In this case, the second model, which refers to the first window, has
the lowest value in cross-validation, so we can select that model and
store its metrics:

``` r
best.window=sel.best.window(mywpls)
#> [1] "The best model is the Window1 model which has range X1100:X1484"
```

We can also plot spectra highlighting the selected window:

``` r
map.best.window(mywpls)
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />


NEWS.md
# windows.pls (development version)

* Initial CRAN submission.


R/cv.wpls.R
#' Cross-validation for segmented spectral regions of the original spectra.
#' @description
#' Computes and stores cross-validation metrics for one of the three possible modes ‘wpls’, ‘epls’, ‘swpls’.
#'
#' @param xblock A matrix containing one spectra for each observation.
#' @param yblock A vector containing the concentration associated to each spectra in the *xblock* matrix.
#' @param windows Parameter used when either ‘wpls’ or ‘ewpls’ is chosen. Points out how many windows the user wants to divide the spectra in.
#' @param window.size Parameter used when ‘swpls’ is chosen. Indicates the width of the window that slides along the spectra.
#' @param increment Parameter used when ‘swpls’ is chosen. Indicates how many steps the window slides forward.
#' @param cv Number of segments used for cross-validation.
#' @param scale logical, asks to perform standardization.
#' @param ncp Maximum number of principal components to be computed for each model.
#' @param mode 'wpls','ewpls' or 'swpls', see **Details** for more.
#' @import ggplot2
#' @import mdatools
#' @import tidyverse
#' @import readr
#' @import stringr
#' @import grDevices
#' @import dplyr
#' @import tidyr
#' @export
#' @returns
#' Returns a list containing:
#'
#' \item{xblock}{Matrix containing spectra used to train the model.}
#' \item{yblock}{Vector containing values of the dependent variable.}
#' \item{cal}{List containing RMSE and R2 of calibratrion.}
#' \item{cv}{List containing RMSE and R2 of cross-validation.}
#' \item{ncp}{Number of components used to compute the model.}
#' \item{scale}{Contains logical condition used for standardization.}
#' \item{cv.segment}{ Number of segments used for cross-validation.}
#'
#' @details NIR and Vis-NIR technologies are used to obtain spectra which might
#' contain helpful information about the content of the samples the user is investigating.
#' Since this method has been combined with multivariate statistical methods, researchers have
#' been questioning the importance of using spectra in its entirety or if it might be a better
#'  solution to divide it in smaller regions which can guarantee higher performance in terms of predictions.
#'  Several methods have been proposed, from selecting only some regions to selecting combinations of those which are performing the best.
#' This function provides three possibilities:
#'
#'1. **‘wpls’**, which stands for *Window PLS*, divides the original spectra into several windows, computes PLS and stores metrics of interest such as RMSE and R2 for calibration and cross-validation both.
#'
#'2. **‘ewpls’**, which stands for *Evolving Window PLS*, divides the original spectra into several windows, but each new window incorporates the previous ones, so that we are comparing smaller windows with the entire spectra.
#'
#'3. **’swpls’**, which stands for *Sliding Window Window PLS*, ,asks the width of the window that will be used to compute the model and the step that the window will make forward in the spectra so that a new model is calculated. In this way the window slides along spectra and computes several models, which will be compared with metrics.
#'
#'
#'This function proposes a simpler version of **iPLS**, that can be found in the **mdatools** package, which divides the spectra in smaller segments and tries to find the combination with the lowest RMSE in cross-validation.
#'
#' @references
#' 1. Chen, J., Yin, Z., Tang, Y. et al. Vis-NIR spectroscopy with moving-window PLS method applied to rapid analysis of whole blood viscosity. Anal Bioanal Chem 409, 2737–2745 (2017).
#'
#' 2. Y.P. Du, Y.Z. Liang, J.H. Jiang, R.J. Berry, Y. Ozaki, Spectral regions selection to improve prediction ability of PLS models by changeable size moving window partial least squares and searching combination moving window partial least squares, Analytica Chimica Acta, Volume 501, Issue 2, 2004, Pages 183-191,
#'
#' 3. **mdatools** package, https://github.com/svkucheryavski/mdatools
#' @examples
#' data(beer)
#' conc=beer[,1]
#' sp=beer[,2:ncol(beer)]
#' names(sp)=convert.names.wl(1100,2250,2)
#' conc=unlist(conc)
#' mywpls=cv.wpls(sp, conc,mode='wpls', windows = 5)

cv.wpls=function(xblock=NULL,
                 yblock=NULL,
                 windows=3,
                 window.size=30,
                 increment=10,
                 cv=10,
                 scale=F,
                 ncp=10,
                 mode='wpls'){
  if( !(mode%in% c('wpls','ewpls','swpls'))){stop('Mode not available')}
  limits=round(seq(1,ncol(xblock),length.out=windows+1))
  size=seq(1,windows,1)
  # initialize rmse matrix
  perf.rmse.cv=matrix(NA,nrow = ncp)
  perf.rmse.cal=matrix(NA,nrow = ncp)
  # initialize r2 matrix
  perf.r2.cv=matrix(NA,nrow = ncp)
  perf.r2.cal=matrix(NA,nrow = ncp)
  # matrix that contains column index
  index=matrix(NA,ncol = 2)
  wl.window=c()
  nwindows=c()
  mode=mode
  for (i in size){
    # define columns
    if(mode=='swpls'){break}
    j=ifelse(mode=='wpls',limits[i],1)
    J=limits[i+1]
    index.temporary=cbind(j,J)
    x.mw.train=xblock[,j:J]
    # compute model
    pls.mw=pls(x=x.mw.train,
               y=yblock,
               cv=cv,
               ncomp=ncp,
               scale=scale)
    # store rmse
    rmse.cv=as.data.frame(t(pls.mw$cvres$rmse))
    rmse.cal=as.data.frame(t(pls.mw$calres$rmse))
    # store r2
    r2.cv=as.data.frame(t(pls.mw$cvres$r2))
    r2.cal=as.data.frame(t(pls.mw$calres$r2))
    # name of the interval
    nwindow=paste('Window',i,sep='')
    nwindows=c(nwindows,nwindow)
    wl.temporary=paste(make.names(names(xblock)[j]),
                       ':',
                       make.names(names(xblock)[J]),
                       sep='')
    # change name of the first column
    names(rmse.cv)=nwindow
    names(rmse.cal)=nwindow
    # same for r2
    names(r2.cv)=nwindow
    names(r2.cal)=nwindow
    # bind cols rmse
    perf.rmse.cv=cbind(perf.rmse.cv,rmse.cv)
    perf.rmse.cal=cbind(perf.rmse.cal,rmse.cal)
    # bind cols r2
    perf.r2.cv=cbind(perf.r2.cv,r2.cv)
    perf.r2.cal=cbind(perf.r2.cal,r2.cal)
    # store windows
    wl.window=c(wl.window,wl.temporary)
    # store indexes
    index=rbind(index,index.temporary)
  }
  ############ swpls ################
  window.size=window.size
  increment=increment
  j.v=NULL
  J.v=NULL
  for (i in 1:ncol(xblock)){
    if(mode%in% c('wpls','ewpls')){break}
    # define columns
    j=ifelse(i==1,1,j.v[i-1]+increment)
    J=ifelse(i==1,1+window.size,J.v[i-1]+increment)
    if(J>ncol(xblock)){J=ncol(xblock)}
    j.v=c(j.v,j)
    J.v=c(J.v,J)
    if(i>1 && J.v[i-1]==ncol(xblock)){break}
    index.temporary=cbind(j,J)
    x.mw.train=xblock[,j:J]
    # compute model
    pls.mw=pls(x=x.mw.train,
               y=yblock,
               cv=cv,
               ncomp=ncp,
               scale=scale)
    # store rmse
    rmse.cv=as.data.frame(t(pls.mw$cvres$rmse))
    rmse.cal=as.data.frame(t(pls.mw$calres$rmse))
    # store r2
    r2.cv=as.data.frame(t(pls.mw$cvres$r2))
    r2.cal=as.data.frame(t(pls.mw$calres$r2))
    # name of the interval
    nwindow=paste('Window',i,sep='')
    nwindows=c(nwindows,nwindow)
    wl.temporary=paste(make.names(names(xblock)[j]),
                       ':',
                       make.names(names(xblock)[J]),
                       sep='')
    # change name of the first column
    names(rmse.cv)=nwindow
    names(rmse.cal)=nwindow
    # same for r2
    names(r2.cv)=nwindow
    names(r2.cal)=nwindow
    # bind cols rmse
    perf.rmse.cv=cbind(perf.rmse.cv,rmse.cv)
    perf.rmse.cal=cbind(perf.rmse.cal,rmse.cal)
    # bind cols r2
    perf.r2.cv=cbind(perf.r2.cv,r2.cv)
    perf.r2.cal=cbind(perf.r2.cal,r2.cal)
    # store windows
    wl.window=c(wl.window,wl.temporary)
    # store indexes
    index=rbind(index,index.temporary)
  }
  index=index[-1,]
  # bind indexes and cols
  window.index=as.data.frame(cbind(nwindows,wl.window,index))
  window.index[nrow(window.index)+1,]=c('Complete',
                                        paste(make.names(names(xblock)[1]),
                                              ':',
                                              make.names(names(xblock)[ncol(xblock)]),
                                              sep=''),
                                        1,
                                        ncol(xblock))
  if(mode=='ewpls'){window.index=window.index[-(nrow(window.index)-1),]}
  # compute entire model
  complete=pls(xblock,
               yblock,
               ncomp=ncp,
               scale=scale,
               cv=cv)
  # for cross validation of the complete model
  perf.rmse.cv[,1]=t(complete$cvres$rmse)
  names(perf.rmse.cv)[1]='Complete'
  # for calibration of the complete model
  perf.rmse.cal[,1]=t(complete$calres$rmse)
  names(perf.rmse.cal)[1]='Complete'
  # for cross validation of the complete model
  perf.r2.cv[,1]=t(complete$cvres$r2)
  names(perf.r2.cv)[1]='Complete'
  # for calibration of the complete model
  perf.r2.cal[,1]=t(complete$calres$r2)
  names(perf.r2.cal)[1]='Complete'
  # store cv results in a list
  metrics.cv=list('rmse.cv'=perf.rmse.cv,'r2.cv'=perf.r2.cv)
  # store cal results in a list
  metrics.cal=list('rmse.cal'=perf.rmse.cal,'r2.cal'=perf.r2.cal)
  ## create object that contains rmse and r2 results
  metrics=list('xblock'=xblock,
               'yblock'=yblock,
               'cal'=metrics.cal,
               'cv'=metrics.cv,
               'windows'=window.index,
               'ncp'=ncp,
               'scale'=scale,
               'cv.segment'=cv)
  return(metrics)
}
